### PR TITLE
Refactor the API for saving the report notification preferences

### PR DESCRIPTION
**Pull request by tgolen**
- [x] Held on https://github.com/Expensify/Web-Expensify/pull/34059
- [x] Held on https://github.com/Expensify/Web-Expensify/pull/34076

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/211600

### Tests
1. Open an #announce workspace chate
2. Click on the chat header to go to the details
3. Go to the settings
4. Change the notification preferences
5. Refresh/reload the app to ensure the setting stuck

- [ ] Verify that no errors appear in the JS console

### QA Steps
Same as above

- [ ] Verify that no errors appear in the JS console

#### Web

https://user-images.githubusercontent.com/1228807/174681699-c518d970-54ed-445c-b01a-33459c78de88.mp4

Adding in a video to show pusher working to update multiple connected clients

https://user-images.githubusercontent.com/1228807/175089042-a5f795b2-938b-41c6-9728-9b8f20ada67f.mp4

#### Mobile Web

https://user-images.githubusercontent.com/1228807/174681716-c87dbf18-4a5a-40b5-96da-1b9f2ba5807e.mp4

#### Desktop

https://user-images.githubusercontent.com/1228807/174681731-22699d08-2120-4218-a23d-d1213830c1ab.mp4

#### iOS

https://user-images.githubusercontent.com/1228807/174681761-940f600a-20ab-4612-ba72-85e9ac40f004.mp4

#### Android

https://user-images.githubusercontent.com/1228807/174681741-9bbe6fef-379c-4086-911e-2e9933f5ca45.mp4

### CODE REVIEW

File: src/libs/actions/Report.js [hunk 1]
```diff
@@ -1314,8 +1314,14 @@ function syncChatAndIOUReports(chatReport, iouReport) {
  * @param {String} notificationPreference
  */
 function updateNotificationPreference(reportID, notificationPreference) {
-    Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {notificationPreference});
-    DeprecatedAPI.Report_UpdateNotificationPreference({reportID, notificationPreference});
+    const optimisticData = [
+        {
+            onyxMethod: 'merge',
```
Comment (marcaaron) [inline src/libs/actions/Report.js:1319]: NAB, I'm adding a const for this in another PR but might be good to add a const for all the `'merge'`

File: src/libs/actions/Report.js [hunk 2]
```diff
@@ -1314,8 +1314,14 @@ function syncChatAndIOUReports(chatReport, iouReport) {
...
+            key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+            value: {notificationPreference},
+        },
+    ];
+    API.write('UpdateReportNotificationPreference', {reportID, notificationPreference}, {optimisticData});
```
Comment (marcaaron) [inline src/libs/actions/Report.js:1324]: I missed this in the design stage, but if this is optimistic without feedback then we still want to rollback on `jsonCode !== 200`

File: src/libs/actions/Report.js (hunk 1 above)
Comment (tgolen) [reply src/libs/actions/Report.js:1319]: Mind linking to the PR so I can add the same constant? It'll be easier to fix any conflicts that way.

**Review by marcaaron — DISMISSED** [review]

**Comment by tgolen** [comment]
I made an update to web-e so that the pusher response would be sent out to all clients. It didn't result in any updates to this branch though.

File: src/libs/actions/Report.js [hunk 3]
```diff
@@ -1311,15 +1311,22 @@ function syncChatAndIOUReports(chatReport, iouReport) {
...
-/**
- * Updates a user's notification preferences for a chat room
- *
- * @param {Number} reportID
- * @param {String} notificationPreference
- */
-function updateNotificationPreference(reportID, notificationPreference) {
-    Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {notificationPreference});
-    DeprecatedAPI.Report_UpdateNotificationPreference({reportID, notificationPreference});
+function updateNotificationPreference(reportID, previousValue, newValue) {
```
Comment (marcaaron) [inline src/libs/actions/Report.js:1314]: we are missing JSDocs for this.

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Review by roryabraham — COMMENTED** [review]
Except for the missing JSDoc LGTM 👍

**Comment by tgolen** [comment]
Added JS docs 👍

File: src/libs/actions/Report.js [hunk 4]
```diff
@@ -1312,14 +1312,26 @@ function syncChatAndIOUReports(chatReport, iouReport) {
 }
 
 /**
- * Updates a user's notification preferences for a chat room
- *
- * @param {Number} reportID
- * @param {String} notificationPreference
+ * @param {String} reportID
```
Comment (roryabraham) [inline src/libs/actions/Report.js:1315]: I think this might be a number, right?

```suggestion
 * @param {Number} reportID
```

**Comment by tgolen** [comment]
updated @roryabraham

**Review by roryabraham — APPROVED** [review]

**Review by marcaaron — APPROVED** [review]

**Comment by marcochavezf** [comment]
Hi guys, QA reported this issue related to the notification preferences https://github.com/Expensify/App/issues/9805, it can't be reproduced consistently but happens when I change the notification preferences and refresh the browser quickly. Thoughts about this issue?

**Comment by marcaaron** [comment]
Explained in this comment -> https://github.com/Expensify/App/issues/9805#issuecomment-1181197421

**Pull request merged**